### PR TITLE
fix(build): use `pull_request_target` to support changes from forked repositories

### PR DESCRIPTION
**Pull request by daniellacosse**
I was mistaken in thinking that pull requests from forked repositories would be treated like regular pull requests - they in fact, are not, for the security reasons we realized independently. Thankfully, we've basically already hardened for this by adding in an environment approval process. All that should need be done now is switch to pull_request_target and checkout based on the head of the pr given - a process [outlined in this article](https://dev.to/petrsvihlik/using-environment-protection-rules-to-secure-secrets-when-building-external-forks-with-pullrequesttarget-hci).

### CODE REVIEW

**Review by jyyi1 — APPROVED** [review]

**Comment by daniellacosse** [comment]
secrets are populated in the test: https://github.com/Jigsaw-Code/outline-client/runs/5219157147?check_suite_focus=true#step:2:13

However, `pull_request_target` has to be on the branch being merged into for this to work. And having this trigger both on `pull_request` and `pull_request_target` fires a bunch of jobs that are immediately cancelled, which github treats as a failure:

Screen Shot 2022-02-16 at 11 12 03

File: .github/workflows/build_and_test_debug.yml
Comment (fortuna) [inline .github/workflows/build_and_test_debug.yml:22]: [Keeping your GitHub Actions and workflows secure Part 1](https://dev.to/petrsvihlik/using-environment-protection-rules-to-secure-secrets-when-building-external-forks-with-pullrequesttarget-hci) says "Combining pull_request_target workflow trigger with an explicit checkout of an untrusted PR is a dangerous practice that may lead to repository compromise."

Comment (fortuna) [reply .github/workflows/build_and_test_debug.yml:22]: Do we really trust the article you shared?
I also don't quite get what the problem is. Is the PR test trying to use the secret in the user's repo? What's wrong with that?

**Review by fortuna — CHANGES_REQUESTED** [review]

File: .github/workflows/build_and_test_debug.yml
Comment (daniellacosse) [reply .github/workflows/build_and_test_debug.yml:22]: Right, so @jyyi1 wants to be able to develop off of his own forked repository. When making a PR from his forked repo, the secrets were withheld from his CI runs by GitHub for security reasons. As mentioned above, I was mistaken about how GitHub handles this. 

This means that anyone wanting to make a PR from their own fork will be unable, as their CI checks will fail due to the missing secrets, even if we approve those deployments. We already have the environment approvals set up, so we should be safe.

Comment (fortuna) [reply .github/workflows/build_and_test_debug.yml:22]: Perhaps we do this for iOS only then, since the other workflows don't need secrets.

Comment (daniellacosse) [reply .github/workflows/build_and_test_debug.yml:22]: Sure, I guess I can split this workflow up into two halves. We just can't have `pull_request` and `pull_request_target` events in the same workflow, otherwise it gets quite ugly - triggering each job twice on each push.

File: .github/workflows/unsafe_build_and_test_debug.yml
Comment (daniellacosse) [reply .github/workflows/unsafe_build_and_test_debug.yml:30]: Yeah I'm aware and am working on that in another PR: https://github.com/Jigsaw-Code/outline-client/pull/1176

The point of this was to unblock people trying to contribute external to the core outline team. If you're fine waiting until I've figured this out, then I'll close this PR

**Pull request closed without merging**
